man/man3/open_image.3
.TH open_image 3  "LPJmL programmers manual"
.SH NAME
open_image \- Opens socket connection to IMAGE model
.SH SYNOPSIS
.nf
\fB#include "lpj.h"

Bool open_image(Config *\fIconfig\fB);\fP

.fi
.SH DESCRIPTION
Function opens communication channel to IMAGE model. Function is only available if LPJmL was compiled with -DIMAGE flag.
The argument of \fBopen_image\fP is:
.TP
.I config
Pointer to LPJmL configuration data. Must be initialized by \fBinitmpiconfig()\fP or \fBinitconfig()\fP and read by \fBreadconfig()\fP.
.SH RETURN VALUE
Upon successful completion FALSE is returned and TRUE in case of error.

.SH AUTHORS

For authors and contributors see AUTHORS file

.SH COPYRIGHT

(C) Potsdam Institute for Climate Impact Research (PIK), see COPYRIGHT file

.SH SEE ALSO
initmpiconfig(3), initconfig(3), close_image(3)


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project follows a [GNU standard version numbering](https://www.gnu.org/prep/standards/html_node/Releases.html#index-version-numbers_002c-for-releases)
of `major.minor.patch` with
- `major` for substantial changes to code functionality,
- `minor` for additions of modules and functionality, and
- `patch` for technical changes and bugfixes without extending the functionality of the code.

- if applicable, categories to include in the CHANGELOG.md are:
  - Added
  - Changed
  - Deprecated
  - Removed
  - Fixed
  - Security

## [Unreleased]

## [5.7.5] - 2023-08-23

### Added

- outputs:
  - IRRIG_STOR
  - RIVERVOL
  - SWC_VOL
 
### Changed:

- unit of output RES_STORAGE changed from hm3 to dm3 to be in line with IRRIG_STOR, RIVERVOL, LAKEVOL, SWC_VOL

### Fixed

- check for land-use fractions to not exceed 100% after scaling with `landfrac`, re-scale to 100%, print warning see [!173](https://gitlab.pik-potsdam.de/lpjml/LPJmL_internal/-/merge_requests/173)
- add turnover before allocation in cultcftstand to prevent sporadic C balance errors when running with intercrops, see[!175](https://gitlab.pik-potsdam.de/lpjml/LPJmL_internal/-/merge_requests/175)
- number of wet days not read if daily precipitation input is used. A warning is additionally printed to set `"random_prec"` to false.
- boolean `"river_routing"` is read before `"with_lakes"` flag to avoid uninitialized variable.


## [5.7.4] - 2023-08-17

### Changed

- changed quadratic soil evaporation function to sigmoid form and a minimum amount of evaporation of 5% of the available energy, following [Sun et al. 2013](http://dx.doi.org/10.1080/17538947.2013.783635), as described in [!168](https://gitlab.pik-potsdam.de/lpjml/LPJmL_internal/-/merge_requests/168)

## [5.7.3] - 2023-08-16

### Fixed

- correct accounting of blue water transpiration on agriculture stands in outputs
- missing variable initialization for bioenergy grass
- landuse change from irrigated to rainfed setaside moved area to wrong target setaside


## [5.7.2] - 2023-08-11

### Changed:

- removed timestep attributes (monthly/annual) in all output descriptions that are not timestep specific

### Added 

- outputs:
  - PFT_WATER_DEMAND: PFT specific water demand
  - RD: dark respiration
  - NDEPOS: total N deposition as sum of NH4 and NO3 deposition

## [5.7.1] - 2023-08-10

### Fixed

- wrong use of `&f` instead of `f`, in `fscanparam.c`. Bug was introduced in version 5.6.26 and caused that parameter files could not be read properly

## [5.7.0] - 2023-08-10

### Added

- land fraction can now be read from file, with a new flag `"landfrac_from_file"` and a new input `"landfrac"`
- If setting `"landfrac_from_file : true"` cell areas in LPJmL are scaled down by land fraction from input file. If land fraction is zero, land fraction is set to `minlandfrac` and a warning is printed.
- new flag `"with_lakes"` defined
- new outputs `"terr_area"`, `"land_area"` and `"lake_area"` defined writing the terrestrial area (land and lakes),  land area and lake area of each cell
- new parameters `"minlandfrac"` and `"tinyfrac"` added to lpjparams.js
- option `-area filename` added to the `printglobal` utility to scale with the cell areas read from the file specified.

### Changed

- Lake fraction and land-use fractions are rescaled by the reciprocal of land fraction. If lake fraction is greater than one, lake fraction is set to one and a warning is printed.
- Lakes can be enabled without river routing by setting `"with_lakes" : true`. Variable `dmass_lake` is always written to restart file.
- lake fraction is read in utility `lpjprint` and printed
- flag `"shuffle_spinup_climate"` only read if `"nspinup"` is non-zero.
- Terrestrial area and harvest fraction are read in utility `printharvest`. Output format changed to CSV.

## [5.6.27] - 2023-08-08

### Removed

- removed large chunks of code in `allocation_daily_crop.c` that was inactive

### Fixed

- fixed output `PFT_VEGC` to not include `bm_inc.nitrogen` for crops in `daily_agriculture.c`
- fixed global flux summation for `flux.soil.carbon` and `flux.soil.nitrogen` (`+=` instead of `=`), bug was introduced in version 5.6.25

## [5.6.26] - 2023-08-07

### changed

- avoid daily leaf turnover for biomass grass and harvest leaf turnover as dead biomass at harvest event instead (reduces litter and increases harvest)
- introduce harvest dates depending on green/brown harvest instead of C threshold for biomass grass
- literature-based C/N ratios and fn_turnover for biomass grass depending on green/brown harvest configuration
- results and literature references can be found in issue [#275](https://gitlab.pik-potsdam.de/lpjml/LPJmL_internal/-/issues/275).

### Added

- introduced `#define` flag to change from green harvest (default) to brown harvest: `#define BMGR_BROWN`
- two options include green harvest (fresh biomass, high N content - usually used for biofuels) and brown harvest (dry biomass, high lignin content, low N content - usually used for combustion)
- brown harvest uses later harvest dates (defined in lpjparam.js) and higher N recovery (as lower fn_turnover defined in pft.js) than green harvest
- the `#define` flag is required to read the correct flag-specific parameter from the lpjparam.js and pft.js, see issue [#311](https://gitlab.pik-potsdam.de/lpjml/LPJmL_internal/-/issues/311)

## [5.6.25] - 2023-07-29

### Added

- New outputs added for N fluxes from timber extraction and for N pools in wood product pools
- New outputs added for `estab_storage` C and N
- Added missing N fluxes, `estab_storage`, and reservoirs to `globalflux` output
- Added new fluxes to man pages

### Fixed 

- Corrected/added units in comments in output.h
- Corrected units in `flux_sum` man page
- Corrected `N2O_DENIT` output, which now no longer also includes `N2O_NIT` values

src/image/open_image.c
/**************************************************************************************/
/**                                                                                \n**/
/**                  o  p  e  n  _  i  m  a  g  e  .  c                            \n**/
/**                                                                                \n**/
/**     extension of LPJ to couple LPJ online with IMAGE                           \n**/
/**     Opens connection to IMAGE model using TDT compliant socket                 \n**/
/**                                                                                \n**/
/** (C) Potsdam Institute for Climate Impact Research (PIK), see COPYRIGHT file    \n**/
/** authors, and contributors see AUTHORS file                                     \n**/
/** This file is part of LPJmL and licensed under GNU AGPL Version 3               \n**/
/** or later. See LICENSE file or go to http://www.gnu.org/licenses/               \n**/
/** Contact: https://github.com/PIK-LPJmL/LPJmL                                    \n**/
/**                                                                                \n**/
/**************************************************************************************/

#include "lpj.h"

#if defined IMAGE && defined COUPLED

Bool open_image(Config *config /**< LPJmL configuration */
               )               /** \return TRUE on error */
{
  char Revision_exe[10]  = "0000";
  char Date_exe[80]      = "2020/05";
  char TimeNow_exe[80]   = "2020/05";
  char URL_exe[80]       = "https://gitlab.pik-potsdam.de";

  int IRevision;
  int IUrl[80];
  int len,i;

  if(isroot(*config))
  {
    /* Establish the TDT connection */
    printf("Connecting IMAGE model...  image_inport %d, wait %d host %s outport %d\n",config->image_inport,config->wait,config->image_host,config->image_outport);
    fflush(stdout);
    config->in=opentdt_socket(config->image_inport,config->wait);
    if(config->in==NULL)
      return TRUE;
#ifdef DEBUG_IMAGE
    printf("opened INTERFACE_to_LPJ\n");

    printf("opening LPJ_to_INTERFACE image_host %s outport %d\n",config->image_host,config->image_outport);
#endif
    config->out=connecttdt_socket(config->image_host,
                                  config->image_outport);
    if(config->out==NULL)
    {
      close_socket(config->in);
      config->in=NULL;
      return TRUE;
    }
#ifdef DEBUG_IMAGE
    printf("opened LPJ_to_INTERFACE\n");
    printf("All connections to IMAGE are established.\n");
    fflush(stdout);
#endif
    printf("LPJ.exe version information\n");
    printf("URL repository: %s \n", URL_exe);
    printf("Revision: %s \n",Revision_exe);
    printf("Date: %s \n",Date_exe);
    printf("BuildTime: %s \n",TimeNow_exe);
    fflush(stdout);
/* SEND revision_lpj and URL_lpj; should be a string in future*/

/* send revision number to interface */
    IRevision = atoi(Revision_exe);
#ifdef DEBUG_IMAGE
    printf("open_image1 %d \n",IRevision);
#endif
    writeint_socket(config->out, &IRevision,1);
#ifdef DEBUG_IMAGE
    printf("open_image2 %d \n",IRevision);
#endif

  /* send revision url */
#ifdef DEBUG_IMAGE
    printf("open_image3  \n"); 
#endif
    len = (int)strlen (URL_exe);
    for(i=0;i<80;i++)
    {
      IUrl[i] = (i<len) ? URL_exe[i] : ' ';
#ifdef DEBUG_IMAGE
      printf("open_image4 %d %d \n",i, IUrl[i] );
#endif
    }
#ifdef DEBUG_IMAGE
    printf("open_image5  \n"); 
#endif
    writeint_socket(config->out, IUrl,80); 
#ifdef DEBUG_IMAGE
    printf("open_image6  \n"); 
#endif

/* end SEND revision_lpj and URL_lpj; should be a string in future*/

  }
  return FALSE;
} /* of 'open_image' */

#endif
